plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformerJsIr.kt

```diff
@@ -0,0 +1,492 @@
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            type,
+            accessorFunction,
+            IrStatementOrigin.LAMBDA
+        )
+    }
+
+    private fun buildSetField(backingField: IrField, ownerClass: IrExpression?, value: IrGetValue): IrSetField {
+        val receiver = if (ownerClass is IrTypeOperatorCall) ownerClass.argument as IrGetValue else ownerClass
+        val fieldSymbol = backingField.symbol
+        return buildSetField(
+            symbol = fieldSymbol,
+            receiver = receiver,
+            value = value
+        )
+    }
+
+    private fun buildGetField(backingField: IrField, ownerClass: IrExpression?): IrGetField {
+        val receiver = if (ownerClass is IrTypeOperatorCall) ownerClass.argument as IrGetValue else ownerClass
+        return buildGetField(backingField.symbol, receiver)
+    }
+
+    private fun IrCall.getPropertyAccessors(parentFunction: IrDeclaration): List<IrExpression> =
+        listOf(buildGetterLambda(parentFunction), buildSetterLambda(parentFunction))
+
+    private fun getRuntimeFunctionSymbol(name: String, type: IrType): IrSimpleFunctionSymbol {
+        val functionName = when (name) {
+            "value.<get-value>" -> "getValue"
+            "value.<set-value>" -> "setValue"
+            else -> name
+        }
+        return referencePackageFunction("kotlin.js", "$ATOMICFU_RUNTIME_FUNCTION_PREDICATE$functionName") {
+            val typeArg = (it.owner.getGetterParameter().type as IrSimpleType).arguments.first()
+            !(typeArg as IrType).isPrimitiveType() || typeArg == type
+        }
+    }
+
+    private fun IrFunction.getGetterParameter() = valueParameters[valueParameters.lastIndex - 1]
+
+    private fun IrCall.isAtomicFactoryFunction(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_CONSTRUCTOR
+    }
+
+    private fun IrCall.isAtomicArrayFactoryFunction(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_ARRAY_FACTORY_FUNCTION
+    }
+
+    private fun IrConstructorCall.isAtomicArrayConstructor(): Boolean {
+        val name = (type as IrSimpleType).classifier.descriptor.name
+        return !name.isSpecial && name.identifier.matches(Regex(ATOMIC_ARRAY_TYPE))
+    }
+
+    private fun IrSymbol.isKotlinxAtomicfuPackage() =
+        this is IrPublicSymbolBase<*> && signature.packageFqName().asString() == AFU_PKG.prettyStr()
+
+    private fun IrType.isAtomicValueType() = belongsTo(ATOMICFU_VALUE_TYPE)
+    private fun IrType.isAtomicArrayType() = belongsTo(ATOMIC_ARRAY_TYPE)
+    private fun IrType.isReentrantLockType() = belongsTo("$AFU_PKG/$LOCKS", REENTRANT_LOCK_TYPE)
+
+    private fun IrType.belongsTo(typeName: String) = belongsTo(AFU_PKG, typeName)
+
+    private fun IrType.belongsTo(packageName: String, typeName: String): Boolean {
+        if (this !is IrSimpleType || classifier !is IrClassPublicSymbolImpl) return false
```

Comment:
`classifier !is IrClassPublicSymbolImpl` -> `classifier.isPublicApi && classifier is IrClassSymbol`